### PR TITLE
Create2 and EIP1271 support

### PR DESCRIPTION
**Pull request by MaximFischuk**

### CODE REVIEW

File: src/main/java/io/zksync/signer/DefaultEthSigner.java [hunk 1]
```diff
@@ -80,8 +80,23 @@ public static DefaultEthSigner fromRawPrivateKey(Web3j web3j, String rawPrivateK
...
+        Credentials credentials = generateCredentialsFromMnemonic(mnemonic, 0);
+        return new DefaultEthSigner(new RawTransactionManager(web3j, credentials), credentials, contractAddress);
+    }
+
+    public static DefaultEthSigner fromMnemonicEIP1271(Web3j web3j, String mnemonic, int accountIndex, String contractAddress) {
+        Credentials credentials = generateCredentialsFromMnemonic(mnemonic, accountIndex);
+        return new DefaultEthSigner(new RawTransactionManager(web3j, credentials), credentials, contractAddress);
+    }
+
+    public static DefaultEthSigner fromRawPrivateKeyEIP1271(Web3j web3j, String rawPrivateKey, String contractAddress) {
```
Comment (StanislavBreadless) [inline src/main/java/io/zksync/signer/DefaultEthSigner.java:93]: How EIP1271 can be from raw private key? These arent usually created from mnemonics either

File: src/test/java/io/zksync/IntegrationTestCreate2ShortFlow.java [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+import io.zksync.signer.ZkSigner;
+import io.zksync.transport.ZkTransactionStatus;
+import io.zksync.transport.receipt.ZkSyncPollingTransactionReceiptProcessor;
+import io.zksync.transport.receipt.ZkSyncTransactionReceiptProcessor;
+import io.zksync.wallet.SignedTransaction;
+import io.zksync.wallet.ZkASyncWallet;
+import io.zksync.provider.AsyncProvider;
+
+@Ignore
+public class IntegrationTestCreate2ShortFlow {
```
[the remark this answers is not in the document]
Comment (MaximFischuk) [reply src/test/java/io/zksync/IntegrationTestCreate2ShortFlow.java:39]: No, this test flow (like other integration tests) used for check correctness execution Create2 transactions on remote server

File: src/main/java/io/zksync/signer/DefaultEthSigner.java (hunk 1 above)
Comment (MaximFischuk) [reply src/main/java/io/zksync/signer/DefaultEthSigner.java:93]: Basically [EIP1271](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-1271.md) describes only interface of smart contract (called ERC1271). Any smart-contract that implements ERC1271 can verify signatures in any required way. In this case library creates Signer that will make signatures using `rawPrivateKey` and verify signed data with ERC1271 compliant contract deployed at `contractAddress`

File: src/test/java/io/zksync/IntegrationTestCreate2ShortFlow.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (StanislavBreadless) [reply src/test/java/io/zksync/IntegrationTestCreate2ShortFlow.java:39]: I meant, are they run in CI? They should

[the remark this answers is not in the document]
Comment (MaximFischuk) [reply src/test/java/io/zksync/IntegrationTestCreate2ShortFlow.java:39]: Usually include any tests with required remote connection is a bad practice. Integration tests in CI pipeline guaranteed will fail in the future (with any issue like insufficient funds, dropped connection, etc.). Guess will be better to implement unit tests with Mock remote J-RPC server for checking correctness all transaction (an other) requests

[the remark this answers is not in the document]
Comment (StanislavBreadless) [reply src/test/java/io/zksync/IntegrationTestCreate2ShortFlow.java:39]: I now that it is not a good idea and that the tests may fail due to insufficient funds or zkSync server issues. Unfortunately, there is no other way for now and we should get some basic guarantees that the code works

File: src/main/java/io/zksync/signer/Create2EthSigner.java [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+import io.zksync.domain.transaction.ZkSyncTransaction;
+import io.zksync.exception.ZkSyncException;
+
+public class Create2EthSigner implements EthSigner<ChangePubKeyCREATE2> {
+    private final TransactionManager transactionManager;
+
+    private ChangePubKeyCREATE2 authData;
+    private String address;
+
+    private Create2EthSigner(String zkSyncAddress, String address, ChangePubKeyCREATE2 create2Data, TransactionManager transactionManager) {
```
Comment (MaximAlien) [inline src/main/java/io/zksync/signer/Create2EthSigner.java:30]: Do we need `zkSyncAddress` parameter here?

Comment (StanislavBreadless) [reply src/main/java/io/zksync/signer/Create2EthSigner.java:30]: Good call! 

@MaximFischuk is `zkSyncAddress` ever different from `address`?

Comment (MaximFischuk) [reply src/main/java/io/zksync/signer/Create2EthSigner.java:30]: This class can be constructed only used factory methods. Also private constructor cannot be called from anywhere except members of this class. Thus this constructor used only as internal fields initializer

Comment (StanislavBreadless) [reply src/main/java/io/zksync/signer/Create2EthSigner.java:30]: I understand, but is it possible that the `zkSyncAddress` is different from `address`? I mean why do we need to initialize these two fields in the first place? It is not critical, but it is weird

Comment (MaximFischuk) [reply src/main/java/io/zksync/signer/Create2EthSigner.java:30]: > I understand, but is it possible that the `zkSyncAddress` is different from `address`?
In this case it's impossible (except some hacks like construct using reflection :-) )

> I mean why do we need to initialize these two fields in the first place?
I guess `zkSyncAddress` parameter of this constructor is useless, I'll remove it. I'll left only ethereum address cause it used for `getAddress` method

File: src/test/java/io/zksync/IntegrationTestCreate2ShortFlow.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (MaximFischuk) [reply src/test/java/io/zksync/IntegrationTestCreate2ShortFlow.java:39]: Added Mock unti test for Create2 auth

**Review by StanislavBreadless — APPROVED** [review]

**Pull request merged**